src/common/ceph_json.h

```diff
@@ -907,98 +1079,54 @@ class JSONFormattable : public ceph::JSONFormatter {
-  long val_long() const;
-  long long val_long_long() const;
-  bool val_bool() const;
-
-  const std::map<std::string, JSONFormattable> object() const {
-    return obj;
-  }
-
-  const std::vector<JSONFormattable>& array() const {
-    return arr;
-  }
-
-  const JSONFormattable& operator[](const std::string& name) const;
-  const JSONFormattable& operator[](size_t index) const;
-
-  JSONFormattable& operator[](const std::string& name);
-  JSONFormattable& operator[](size_t index);
-
-  operator std::string() const {
-    return value.str;
-  }
-
-  explicit operator int() const {
-    return val_int();
-  }
-
-  explicit operator long() const {
-    return val_long();
-  }
-
-  explicit operator long long() const {
-    return val_long_long();
-  }
-
-  explicit operator bool() const {
-    return val_bool();
-  }
-
-  template<class T>
-  T operator[](const std::string& name) const {
-    return this->operator[](name)(T());
-  }
-
-  template<class T>
-  T operator[](const std::string& name) {
-    return this->operator[](name)(T());
-  }
-
-  std::string operator ()(const char *def_val) const {
-    return def(std::string(def_val));
-  }
-
-  int operator()(int def_val) const {
-    return def(def_val);
-  }
-
-  bool operator()(bool def_val) const {
-    return def(def_val);
-  }
-
-  bool exists(const std::string& name) const;
-  bool exists(size_t index) const;
-
-  std::string def(const std::string& def_val) const;
-  int def(int def_val) const;
-  bool def(bool def_val) const;
+};
+WRITE_CLASS_ENCODER(JSONFormattable)
 
-  bool find(const std::string& name, std::string *val) const;
+static inline JSONFormattable default_formattable;
 
-  std::string get(const std::string& name, const std::string& def_val) const;
+inline JSONFormattable& JSONFormattable::operator[](const std::string& name) {
+	if(const auto i = obj.find(name); end(obj) != i)
+	 return i->second;
+	
+	return default_formattable;
+}
 
-  int get_int(const std::string& name, int def_val) const;
-  bool get_bool(const std::string& name, bool def_val) const;
+inline const JSONFormattable& JSONFormattable::operator[](const std::string& name) const {
+	return const_cast<JSONFormattable *>(this)->operator[](name);
+}
 
-  int set(const std::string& name, const std::string& val);
-  int erase(const std::string& name);
+inline JSONFormattable& JSONFormattable::operator[](size_t index) {
+	return index >= arr.size() ? default_formattable : arr[index];
+}
 
-  void derive_from(const JSONFormattable& jf);
+inline const JSONFormattable& JSONFormattable::operator[](size_t index) const {
+	return const_cast<JSONFormattable *>(this)->operator[](index);
+}
 
-  void encode_json(const char *name, ceph::Formatter *f) const;
+inline void encode_json(const char *name, const JSONFormattable& v, Formatter *f)
+{
+  v.encode_json(name, f);
+}
 
-  bool is_array() const {
-    return (type == FMT_ARRAY);
+inline void JSONFormattable::encode_json(const char *name, Formatter *f) const
```

Comment:
Honestly I've always preferred defining these inside the class and only having 'out-of-line' definitions in the `.cc`. file, though I know that's a matter of taste.